keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+### Algorithm
+
+The pseudo algorithms to support each user story are described as below.
+
+#### NodeAffinity
+
+```bash
+for each candidate node; do
+    if "EvenSpreading" is enabled; then
+        count number of matching pods on the topology domain this node belongs to
+        if matching num - minMatching num < defined MaxSkew; then
+            add it to candidate list
+        fi
+    else
+        keep current logic as is
+    fi
+done
+```
+
+For example, on a 2-zone cluster, suppose an application is specified with
+NodeAffinity "app in [zone1, zone2]".
+
+- If "MaxSkew" is 1, rollout of its replicas can be 1/0 => 1/1 => 2/1. It can't
+  be 2/0 or 3/1.
+- If "MaxSkew" is 2, rollout of its replicas can be 1/0 => 1/1 or 2/0.
+
+This algorithm works for both Design A<sup>1</sup> and B.
+
+<sup>1</sup> As NodeAffinityTerm doens't have info like
+podSelector/namespaces/topologyKey, to make Design A work, those info need to be
+inferred based on NodeAffinity spec.
+
+#### PodAffinity
+
+```bash
+for each candidate node; do
+    if "EvenSpreading" is enabled; then
+        count number of matching pods on the topology domain this node belongs to
+        if matching num - minMatching num < defined MaxSkew; then
+            add it to candidate list
+        fi
+    else
+        keep current logic as is
+    fi
+done
+```
+
+For example, in a 3-zones cluster, the matching pods spread as 2/1/0, then
+
+- if "MaxSkew" is 1, incoming pod can only be deployed onto zone3 - i.e. 2/1/1
+- if "MaxSkew" is 2, incoming pod can be deployed onto zone2 or zone3 - i.e.
+  2/2/0 or 2/1/1
+
+This algorithm works for both Design A and B.
+
+#### PodAntiAffinity
+
+Design A provides a way to trigger "toleration" mode to broaden semantics of
+PodAntiAffinity so as to provide a solution for Story 3. The algorithm is
+roughly described as below:
+
+```bash
+# need to have a structure to know if there is at least one qualified candidate
+existCandidate = true if there is one qualified candidate globally else false
+if not existCandidate and "EvenSpreading" is disabled; then
+    return
+fi
+for each candidate node; do
+    if existCandidate; then
+        add it to candidate list if this node is qualified
+    elif "EvenSpreading" is enabled; then
+        count number of miss-matching pods on the topology domain this node belongs to
+        if misMatching# - minMisMatching# < defined MaxSkew; then
+            add it to candidate list
```

Comment:
what is this candidate list ? I am assuming this is happening in the context of the scheduler selecting a node for pod being scheduled ? so the candidate list is the node where the current pod needs to be scheduled ?